src/flint/types/nmod_poly.pyx

```diff
@@ -355,17 +471,85 @@ cdef class nmod_poly(flint_poly):
     def __rmod__(s, t):
         return divmod(t, s)[1]      # XXX
 
-    def __pow__(nmod_poly self, exp, mod):
+    def __pow__(nmod_poly self, exp, mod=None):
         cdef nmod_poly res
         if mod is not None:
-            raise NotImplementedError("nmod_poly modular exponentiation")
+            return self.powmod(exp, mod)
         if exp < 0:
             raise ValueError("negative exponent")
         res = nmod_poly.__new__(nmod_poly)
         nmod_poly_init_preinv(res.val, (<nmod_poly>self).val.mod.n, (<nmod_poly>self).val.mod.ninv)
         nmod_poly_pow(res.val, self.val, <ulong>exp)
         return res
 
+    def powmod(self, e, modulus, mod_rev_inv=None):
+        """
+        Returns ``self`` raised to the power ``e`` modulo ``modulus``:
+        :math:`f^e \mod g`/
+
+        ``mod_rev_inv`` is the inverse of the reverse of the modulus,
+        precomputing it and passing it to ``powmod()`` can optimise
+        powering of polynomials with large exponents.
+
+            >>> x = nmod_poly([0,1], 163)
+            >>> f = 30*x**6 + 104*x**5 + 76*x**4 + 33*x**3 + 70*x**2 + 44*x + 65
+            >>> g = 43*x**6 + 91*x**5 + 77*x**4 + 113*x**3 + 71*x**2 + 132*x + 60
+            >>> mod = x**4 + 93*x**3 + 78*x**2 + 72*x + 149
+            >>> 
+            >>> f.powmod(123, mod) 
+            3*x^3 + 25*x^2 + 115*x + 161
+            >>> f.powmod(2**64, mod)
+            52*x^3 + 96*x^2 + 136*x + 9
+            >>> mod_rev_inv = mod.reverse().inverse_series_trunc(4)
+            >>> f.powmod(2**64, mod, mod_rev_inv)
+            52*x^3 + 96*x^2 + 136*x + 9
+        """
+        cdef nmod_poly res
+
+        if e < 0:
+            raise ValueError("Exponent must be non-negative")
```

Comment:
Would it make sense to use `inverse_mod` here?

Or maybe the caller should just do that if needed.

Reply:
Yeah good question. I don't think we invert where e is negative elsewhere, but this is a reasonable thing to do.

Reply:
Inverting does bring additional complications though because I think then we need the modulus to be irreducible.